p4-16/spec/P4-16-spec.mdk

```diff
@@ -4378,6 +4380,26 @@ select (p.tcp.port) {
 The expression `16w0 &&& 16w0xFC00` describes the set of 16-bit
 values whose most significant six bits are zero.
 
+Some targets may support parser value set, see Section [#sec-value-set]. Given
+a type `T` for the type parameter of the value set, the type of the value set
+is `set<T>`. The type of the value set must match to the type of all other
+`keysetExpression` in the same `select` expression. If there is a mismatch, the
+compiler must raise an error.
+
+For example, to allow control plane API to specify TCP reserved ports at
+runtime, one could write:
+
+~ Begin P4Example
+value_set<bit<16>>(4) pvs;
```

Comment:
This definition is still not enough to generate a good API, because there is no way to name the fields, the values for which are going to be passed when value_set entries are going to be populated. This is especially important in case value_set is used to match against a tuple. Neither it is clear whether the values passed should be single or value/mask pair, range or anything else.

Reply:
In this example there is just one field.
We can allow a value_set take a struct type, and in that case the fields can be named like the struct fields.

Reply:
@vgurevich : this spec does not specify anything about the control-plane API. The way values are specified should be part of that other document.